### ISSUE TITLE
Waveform Generator + Speaker Driver Module - Prototyping

### ISSUE THREAD

**Issue by digitaldanny** [digitaldanny/GestureControl#38]
**Describe the solution you'd like**
- Recognize when the glove is plugged into the charger.
- Select between 4 notes automatically without assistance from the MCU. This will likely use a 555 timer, some logic gates, a 2049 up/down counter, and a 4051 analog demux to select between resistors that control a different 555's frequency. This logic should count **up** for plugging in the device or detecting a full charge, and it should count **down** when the device is unplugged.
- Generate a waveform using a 555 timer. The circuit described above will provide the resistor to select frequency.
- Volume controller. Negative feedback op amp with a potentiometer for the feedback resistance allowing voltage range of 0 - 3.3V.
- Class AB power amplifier for allowing low impedance speaker to draw a lot of current. This will be based on an npn and a pnp BJT.

**Additional context**
- Research for this issue was completed in #33.
- Circuitry to detect if the battery is fully charged will be completed in another issue. During prototyping of this hardware, I will use an extra DAD board pin set to 3.3V or 0V to determine say if the battery is "charged" or "not charged."

**Comment by digitaldanny** [comment]
Some truth tables that need the logic gates to be defined.

**4029 counter outputs to 4051 demux outputs logic**
The demux outputs control which resistor values are chosen to select the 555 timer frequency.

| C3 | C2 | C1 | C0 | - | D3 | D2 | D1 | D0 |
|---|---|---|---|---|---|---|---|---|
| * | * | 0 | 0 | - | 0 | 0 | 0 | 1 |
| * | * | 0 | 1 | - | 0 | 0 | 1 | 0 |
| * | * | 1 | 0 | - | 0 | 1 | 0 | 0 |
| * | * | 1 | 1 | - | 1 | 0 | 0 | 0 |

**Circuit inputs to 4029 counter direction**
- Controls whether the notes are played in forward or reverse direction.
- This should be split into 2 separate circuits that are combined in a later stage.

| Connected (Active Low) | Battery Charged | - | Counter Direction | Description |
|---|---|---|---|---|
| Falling edge | 0 | - | Up | Connecting but NOT charged. | 
| 0 | Rising edge | - | Up | Fully charged and connected. | 
| Rising edge | 0 | - | Down | Disconnecting. | 

Circuit 1
| Connected (Active Low) | Battery Charged | - | Counter Direction | Description |
|---|---|---|---|---|
| 0 | Rising edge | - | Up | Fully charged and connected. | 

Circuit 2
| Connected (Active Low) | Battery Charged | - | Counter Direction | Description |
|---|---|---|---|---|
| Falling edge | 0 | - | Up | Connecting but NOT charged. | 
| Rising edge | 0 | - | Down | Disconnecting. | 

**Todo**
- Logic for resetting the counter.
- Logic for enabling clock for 4 clock cycles.
- Create a digital hardware diagram for the logic.

**Comment by digitaldanny** [comment]
[Monostable Multivibrator Mode (555 timer)](https://www.electronicshub.org/monostable-multivibrator-using-555-timer/): Some notes about the 555 timer's one-shot mode that could be useful.

- Monostable multivibrator mode ("one-shot" pulse generator): When a trigger input is applied, a pulse is produced at the output and returns back to the stable state after a time interval. The duration that the timer is high depends on an RC timing circuit.
- This can be useful for timing divider circuits. If the 555 period high is set to be slightly longer than the input pulse period, input pulses can be skipped. This acts like a clock divider. 
- I believe this can be useful for this project in combination with another 555 timer and an AND gate to limit the output to 4 clock pulses.

![Monostable Multivibrator Mode](https://user-images.githubusercontent.com/40513675/106670216-be6a8700-657a-11eb-9b46-5739e0350366.JPG)

**Comment by digitaldanny** [comment]
Some additional notes on the 555 timer modes..

- Astable mode: Outputs oscillating pulse waveform. Tunable output frequency and duty cycle. 
- Bistable mode: Toggle output pin between high and low states depending on the state of the two inputs. Acts like a flip flop.

**Comment by digitaldanny** [comment]
[Rising and Falling Edge Detector - YouTube](https://www.youtube.com/watch?v=cqZLJXYXurg): This video shows how to make rising/falling edge detectors with a single NPN transistor. This can be used for detecting if the device is connected or disconnected.

**Comment by digitaldanny** [comment]
I've attached a high level diagram of how this circuit will work below. I've also listed an explanation of the steps starting from the bottom when the glove PCB is connected/disconnected.

- Charger and Glove PCBs will each have 2 pins for detecting glove connection without requiring power from the glove PCB. The glove PCB shorts the two pins together. If the glove is disconnected, the pull-up resistor will bring the pin high. When the glove is connected, the pin is shorted to ground, bringing the pin low. NOTE - the diagram incorrectly shows 2 pull-up resistors, but only 1 will be used in the prototype.
- There will be 2 npn transistor based circuits for detecting positive edges and negative edges to notify the rest of the circuit that a connection/disconnection has just happened.
- The DFF is used to control whether the 4029 counter will count up or down. This is because I want the tune to be played forward or backwards depending on if the glove is connected or disconnected.
- The edge detectors will trigger a monostable 555 timer, which will enable an astable mode 555 timer long enough to output 4 pulses. 
- The 4029 counter will count from 0-3 or 3-0 depending on if it is in up or down mode. 
- The 4051 demux decodes the 0-3 binary values and selects a resistor value to be used in the next stage's 555 timer for frequency selection. This stage may also need some BJTs or MOSFETS - I still need to experiment with this on the breadboard.
- The 3rd 555 timer and RC circuit will output a triangle wave at the frequency selected in the previous stage. This output is what will be played through the speaker, so the resistors need to be carefully tuned to notes in the song. NOTE - the drawing should probably also have a connection from the monostable 555's output to this stage's 555 timer threshold.
- The volume controller will amplify the signal from 0V - 3.3V.
- The AB amplifier will keep a 1:1 voltage gain but allows for more current to be drawn from the power supply rather than the volume control amplifier. Output will be fed into the speaker.

![Mid level HW diagram](https://user-images.githubusercontent.com/40513675/107477048-1558f900-6b45-11eb-8bcc-b558f862f3b6.PNG)

**Comment by digitaldanny** [comment]
HW Prototype Parts List

Part | Quantity | Ordered?
------
- [NE555 Timer](https://www.digikey.com/en/products/detail/NE555P/296-NE555P-ND/277057?itemSeq=353228467) | 3 | yes
- [CD4029 Up/Down Counter](https://www.digikey.com/en/products/detail/CD4029BE/296-2046-5-ND/67275?itemSeq=353229961) | 1 | yes
- [4059 Analog Mux/Demux](https://www.digikey.com/en/products/detail/CD74HCT4051E/296-2119-5-ND/38597?itemSeq=353230265) | 1 | yes
- [D Flip-Flop](https://www.digikey.com/en/products/detail/M74HC174B1R/497-1792-5-ND/591885?itemSeq=353231103) | 1 | yes
- [PN2222A NPN BJT](https://www.digikey.com/en/products/detail/PN2222A+PBFREE/1514-PN2222APBFREE-ND/4806937?itemSeq=353249958) | 4+ | yes
- [PN2907ABU PNP BJT](https://www.digikey.com/en/products/detail/PN2907ABU/PN2907ABUFS-ND/6538?itemSeq=353250197) | 1 | yes
- 10 kOhm potentiometer | 3+, more would be good for testing | no
- [8Ohm 1A Speaker (550 Hz - 18k Hz)](https://www.digikey.com/en/products/detail/pui-audio/ASE03008MR-LW150-R/4700907?utm_adgroup=Speakers&utm_source=google&utm_medium=cpc&utm_campaign=Shopping_Product_Audio%20Products&utm_term=&utm_content=Speakers&gclid=CjwKCAiAjeSABhAPEiwAqfxURZxg6c8iRY9oyTtR06iciogIJI-IK7ABkPDLswx1KUCg35I3fzJnfBoC3iAQAvD_BwE) | 1 | yes

These parts may not be needed, but I'm ordering them just in case.
- [2N7000 N-channel MOSFETs (may be needed)](https://www.digikey.com/en/products/detail/on-semiconductor/2N7000/244278)
- [BS250P P-channel MOSFETs (may be needed)](https://www.digikey.com/en/products/detail/diodes-incorporated/BS250P/92630) 
- [1N4148 General Purpose Diode](https://www.digikey.com/en/products/detail/on-semiconductor/1N4148/458603)
- [1N4007 Rectifier Diode](https://www.digikey.com/en/products/detail/nte-electronics-inc/1N4007/11645794)

**Comment by digitaldanny** [comment]
Update for 02/14.

In Progress:
- Worked on the Class AB amp. 
  - Not sure why, but my output signal sometimes goes rail-to-rail and sometimes only outputs the top half of the input signal. 
  - Most likely, I just have something wired wrong.
- The waveform generating 555 is set up in astable mode. May need to adjust Rb once I determine the resistances needed to produce notes A, B, C, D (or whatever notes I end up choosing).
- Working on selecting between resistance values using the demux. 
  - The mux was not able to output enough current to act as the 5V source connecting Ra with the 555 timer. Ended up using 4 N-Ch MOSFETs to supply this voltage. The mux outputs control the FET bases. 
  - Also had trouble using the FETs to supply the 5V output until I realized that I was using 12kOhm resistors in the FET circuits. This means the current was 5/12k = 0.400 mA, which seems to be not enough power. Using 100 Ohm resistors in place seems to fix the issues.
  - Still seeing problems getting a **single** resistor to be enabled at a time. I think it may have something to do with the MOSFET Vouts switching between 5V and 0V. The resistors end up being added in parallel anyways, which is resulting in strange behaviors.

Todo:
- <s>Use the up/down counter to drive the mux selects.</s>
- <s>Set up the other 2 555 timers to control how long a note is played</s> and how long the system is enabled.
- Implement the logic circuit to control the up/down direction.

**Comment by digitaldanny** [comment]
These are the notes I will tune the resistors to.

| Note | Freq (Hz) | Ra |
| --- | --- | --- |
| C4 | 261.63 | --- |
| E4 | 329.63 | --- |
| G4 | 392.00 | --- |
| C5 | 523.25 | --- |

Assuming Rb = ? and C = ?.

**Comment by digitaldanny** [comment]
Update for 02/14

Completed:
- Finished selecting resistance values with the demux. The simple solution was to put diodes between each resistor and the demux output lines. When a demux output was set to 0, the diode is in reverse bias so there is not current flowing through the connected resistor. When set to Vcc, the diode is in forward bias and allows the specific resistor value to be used for Ra.
- Added the 4029 up/down counter and tested that this could be used to control the demux. Just waiting for the up/down logic to be added to be 100% done.
- Added the 555 to control how long a note is played. The resistors/capacitor are currently set to 0.5Hz (changes every 2 seconds), but I will modify these values later to be around 2Hz.

**Comment by digitaldanny** [comment]
Update on the edge detector design. I was originally planning on building this with npn transistors, but [this StackExchange comment](https://electronics.stackexchange.com/a/270908) shows a clever solution to building a rising _and_ falling edge detector using an RC circuit and XOR gate. The XOR inputs both come from the input; however, one of the inputs are delayed by the RC circuit. There is a brief period of time when the input voltage switches from high-to-low or low-to-high when the inputs will be different values. This is what creates the output pulse.

I'm using this circuit with an [NPN transistor inverter](http://sullystationtechnologies.com/npninverter.html#:~:text=The%20output%20of%20an%20inverter,the%20output%20is%20logical%200.&text=The%20transistor%20inverter%20is%20implemented,junction%20transistors%20could%20be%20substituted.) to combine the positive-edge detector, negative-edge detector, and OR gate into a single circuit. 

Since I do not have any XOR ICs, I will just use 4 NAND gates as shown in [this link](https://www.researchgate.net/figure/Constructing-planar-crossover-gadgets-with-NAND-gates-a-XOR-symbol-b-NAND-symbol_fig1_335233129).

**Comment by digitaldanny** [comment]
Current issues with the circuit..

1.) The beat 555 timer has an offset before the capacitor charges up enough to switch voltages.

![Beat Timer Delay](https://user-images.githubusercontent.com/40513675/108175276-93ca1380-70ce-11eb-8623-19babc1f84a1.PNG)

2.) The 4 resistance values need to be modified for the timer to output the frequencies required.
3.) The duty cycle is closer to 70% than 50% for this timer as well. May need to increase the size of Rb.
4.) There is ***sometimes*** a delay before the up/down dir signal from the DFF sets the direction when the connected pin is pulled high. It seems like this is a race between the up/down dir signal and the enable signal from the monostable 555 timer.
5.) The volume is extremely quiet even at its max. How do I improve this? 
6.) <s>Need to force the 4029 initial state to 0. Right now, it can initialize to anything.</s>

**Comment by digitaldanny** [comment]
> Need to force the 4029 initial state to 0. Right now, it can initialize to anything.

Fixed using the suggestion on [this StackExchange comment](https://electronics.stackexchange.com/a/123103). I used a 100nF cap to enable the Preset Enable during power on and a 100k resistor to discharge the capacitor immediately after. This results in a consistent initial state of 0 on the 4029.

**Comment by digitaldanny** [comment]
Altium schematic completed for this hardware. Currently focused on the battery charger schematic / prototype before completing the charger PCB placement.

**Comment by digitaldanny** [comment]
No work completed on this since issue was closed. Just leaving snapshots of current schematic for later reference. 

From the bottom up..

**Glove Connection Detector**
![D) Glove Connection Detectors](https://user-images.githubusercontent.com/40513675/109099998-056b1880-76f2-11eb-9c6b-cbfdc2aacc98.PNG)

**Tune Up/Down Direction Controller**
![E) Up Down Dir Controller](https://user-images.githubusercontent.com/40513675/109100010-0ac86300-76f2-11eb-8a14-1bed215d98b1.PNG)

**Frequency Controller**
![C) Frequency Controller Logic](https://user-images.githubusercontent.com/40513675/109100017-0ef48080-76f2-11eb-8257-848727db4938.PNG)

**Note Waveform Generator**
![G) Note Generator](https://user-images.githubusercontent.com/40513675/109100021-13b93480-76f2-11eb-9e07-78a711f8b34f.PNG)

**Volume Amplifier**
![F) Volume Amp](https://user-images.githubusercontent.com/40513675/109100026-161b8e80-76f2-11eb-803c-6fabf963ddf7.PNG)

**Class AB Amplifier**
![B) Class AB Amp](https://user-images.githubusercontent.com/40513675/109100036-1ae04280-76f2-11eb-98b7-123dc6a16db8.PNG)
